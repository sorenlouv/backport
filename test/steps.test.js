const axios = require('axios');
const inquirer = require('inquirer');
const nock = require('nock');
const os = require('os');
const httpAdapter = require('axios/lib/adapters/http');

const commitsMock = require('./mocks/commits.json');

const initSteps = require('../src/cli/steps');
const github = require('../src/lib/github');
const rpc = require('../src/lib/rpc');

describe('run through steps', () => {
  beforeEach(() => {
    const owner = 'elastic';
    const repoName = 'kibana';
    const upstream = `${owner}/${repoName}`;
    axios.defaults.host = 'http://localhost';
    axios.defaults.adapter = httpAdapter;

    os.homedir = jest.fn(() => '/homefolder');
    rpc.exec = jest.fn().mockReturnValue(Promise.resolve());
    rpc.writeFile = jest.fn().mockReturnValue(Promise.resolve());
    rpc.mkdirp = jest.fn().mockReturnValue(Promise.resolve());
    rpc.statSync = jest.fn(() => ({ mode: 33152 }));

    github.getCommits = jest.fn(github.getCommits);
    github.createPullRequest = jest.fn(github.createPullRequest);

    inquirer.prompt = jest
      .fn()
      .mockReturnValueOnce(
        Promise.resolve({
          promptResult: {
            message: 'myCommitMessage',
            sha: 'mySha'
          }
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          promptResult: '6.2'
        })
      );

    nock('https://api.github.com')
      .get(`/repos/${owner}/${repoName}/commits`)
      .query({ author: 'sqren', per_page: '5', access_token: 'myAccessToken' })
      .reply(200, commitsMock);

    nock('https://api.github.com')
      .get(`/search/issues`)
      .query({
        q: 'repo:elastic/kibana mySha',
        access_token: 'myAccessToken'
      })
      .reply(200, {
        items: [
          {
            number: 'myPullRequest'
          }
        ]
      });

    nock('https://api.github.com')
      .post(`/repos/${owner}/${repoName}/pulls`)
      .query({ access_token: 'myAccessToken' })
      .reply(200, {
        html_url: 'myHtmlUrl'
      });

    return initSteps({
      username: 'sqren',
      accessToken: 'myAccessToken',
      upstream,
      branches: ['6.x', '6.0', '5.6', '5.5', '5.4'],
      own: true
    });
  });

  it('getCommit should be called with correct args', () => {
    expect(github.getCommits).toHaveBeenCalledWith(
      'elastic',
      'kibana',
      'sqren'
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(github.createPullRequest).toHaveBeenCalledWith('elastic', 'kibana', {
      base: '6.2',
      body: `Backports the following commits to 6.2:\n - myCommitMessage (#myPullRequest)`,
      head: 'sqren:backport/6.2/pr-myPullRequest',
      title: '[6.2] myCommitMessage'
    });
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(rpc.exec.mock.calls).toMatchSnapshot();
  });
});
