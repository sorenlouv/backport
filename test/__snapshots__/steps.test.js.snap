// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run through steps exec should be called with correct args 1`] = `
Array [
  Array [
    "git clone git@github.com:elastic/kibana",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic",
    },
  ],
  Array [
    "git remote add sqren git@github.com:sqren/kibana",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch origin 6.2 && git branch backport/6.2/pr-myPullRequest origin/6.2 --force && git checkout backport/6.2/pr-myPullRequest ",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.2/pr-myPullRequest --force",
    Object {
      "cwd": "/homefolder/.backport/repositories/elastic/kibana",
    },
  ],
]
`;

exports[`run through steps prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "message": "Update readme: Make ui-framework url a link",
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "message": "Update react 15.6.1 (#13672)",
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "message": "Mock out static files when running in Jest (#13315)",
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "promptResult",
        "pageSize": 3,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select branch to backport to",
        "name": "promptResult",
        "type": "list",
      },
    ],
  ],
]
`;
